ast.go

```diff
@@ -3770,42 +3771,78 @@ func (e *Wildcard) String() string {
 
 // CloneExpr returns a deep copy of the expression.
 func CloneExpr(expr Expr) Expr {
+	// Call CloneExprWithConstraints but with basically no constraints -- a
+	// background context which will carry no cancellation information and 0
+	// maximum depth which means infinite recursion.
+	return func() Expr { e, _ := CloneExprWithConstraints(context.Background(), 0, expr); return e }()
+}
+
+func CloneExprWithConstraints(ctx context.Context, maxDepth int, expr Expr) (Expr, error) {
+	return cloneExprWithConstraints(ctx, 0, maxDepth, expr)
+}
+
+func cloneExprWithConstraints(ctx context.Context, depth int, maxDepth int, expr Expr) (Expr, error) {
 	if expr == nil {
-		return nil
+		return nil, nil
 	}
+
+	if maxDepth > 0 && depth >= maxDepth {
+		return nil, fmt.Errorf("max recursion depth of %d exceeded while cloning expression %q", maxDepth, expr)
+	}
+
+	if err := ctx.Err(); err != nil {
```

Comment:
Are we sure this is safe? It seems like it could potentially cause a race condition.

Reply:
the stored error value is wrapped in a mutex in the context.Context type.